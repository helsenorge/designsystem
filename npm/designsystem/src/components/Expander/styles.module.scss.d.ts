export type Styles = {
  button: string;
  button__arrow: string;
  'button__arrow--both-icons': string;
  'button__both-icons': string;
  'button__left-icon': string;
  'button__only-icon': string;
  'button__right-icon': string;
  'button__right-unicode-arrow': string;
  button__text: string;
  'button__text--centered': string;
  'button__text--ellipsis': string;
  'button--arrow': string;
  'button--arrow--both-icons': string;
  'button--borderless': string;
  'button--both-icons': string;
  'button--destructive': string;
  'button--large': string;
  'button--left-icon': string;
  'button--normal': string;
  'button--on-dark': string;
  'button--only-icon': string;
  'button--outline': string;
  'button--right-icon': string;
  'button-wrapper': string;
  'button-wrapper--fluid': string;
  diagonal: string;
  diagonal__line: string;
  'diagonal--on-dark': string;
  expander: string;
  expander__button: string;
  expander__button__text: string;
  'expander__button--expanded': string;
  'expander__button--sticky': string;
  expander__content: string;
  'expander__content--banana': string;
  'expander__content--blueberry': string;
  'expander__content--cherry': string;
  'expander__content--expanded': string;
  'expander__content--icon': string;
  'expander__content--kiwi': string;
  'expander__content--large': string;
  'expander__content--nested-line': string;
  'expander__content--neutral': string;
  'expander__content--plum': string;
  'expander__content--small': string;
  'expander__content--white': string;
  expander__icon: string;
  'expander__icon--left': string;
  'expander__icon--right': string;
  expander__trigger: string;
  'expander__trigger--banana': string;
  'expander__trigger--blueberry': string;
  'expander__trigger--cherry': string;
  'expander__trigger--expanded': string;
  'expander__trigger--icon': string;
  'expander__trigger--kiwi': string;
  'expander__trigger--large': string;
  'expander__trigger--neutral': string;
  'expander__trigger--plum': string;
  'expander__trigger--sticky': string;
  'expander__trigger--white': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
