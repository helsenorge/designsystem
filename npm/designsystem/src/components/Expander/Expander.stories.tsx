import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Expander, { ExpanderSize } from './Expander';
import LawBook from '../Icons/LawBook';
import Loader from '../Loader';
import { Overlay } from '../Loader/Loader';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Expander',
  component: Expander,
  parameters: {
    docs: {
      description: {
        component: 'Expander skjuler detaljinformasjon når den ikke trengs, og gjør den lett tilgjengelig i kontekst når den trengs.',
      },
    },
  },
  args: {
    children:
      'Skjoldbruskkjertelen ligger ved strupehodet på halsen. Den produserer stoffskiftehormonene tyroksin T4 og trijodotyronin T3, som blir ført med blodet til cellene i kroppen. Disse hormonene kontrollerer stoffskiftet.',
    title: 'Hva skjer i kroppen',
    size: ExpanderSize.small,
    color: 'blueberry',
    noNestedLine: false,
    expanded: false,
    renderChildrenWhenClosed: false,
    sticky: true,
  },
  argTypes: {
    children: {
      control: 'text',
    },
    title: {
      control: 'text',
    },
    size: {
      control: 'select',
      options: ExpanderSize,
    },
    color: {
      control: 'select',
      options: ['banana', 'blueberry', 'cherry', 'kiwi', 'neutral', 'plum', 'white'],
    },
    noNestedLine: {
      control: 'boolean',
    },
    expanded: {
      control: 'boolean',
    },
    renderChildrenWhenClosed: {
      control: 'boolean',
    },
    sticky: {
      control: 'boolean',
    },
  },
} satisfies Meta<typeof Expander>;

export default meta;

type Story = StoryObj<typeof meta>;

export const DefaultSmall: Story = {
  render: args => (
    <>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <Expander {...args} />
      <Expander {...args} />
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
      <p>
        {
          'I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.I selvhjelpsprogrammet Co-mestring lærer du om milde til moderate psykiske reaksjoner på korona-situasjonen, og du vil få kunnskap og verktøy for å håndtere dette. Du får daglig lese tekst og gjøre øvelser som du kan bruke for å øve på å håndtere hverdagen.'
        }
      </p>
    </>
  ),
};

export const Large: Story = {
  args: {
    size: ExpanderSize.large,
  },
  render: args => <Expander {...args}>{args.children}</Expander>,
};

export const WithIcon: Story = {
  args: {
    svgIcon: LawBook,
    size: ExpanderSize.large,
  },
  render: args => <Expander {...args}>{args.children}</Expander>,
};

export const WithLoader: Story = {
  render: args => (
    <Expander {...args}>
      <Loader overlay={Overlay.parent} />
    </Expander>
  ),
};

export const WithCallback: Story = {
  args: {
    onExpand: isExpanded => console.log(isExpanded),
  },
  render: args => (
    <Expander {...args}>
      <p>Sjekk nettleserkonsollen</p>
    </Expander>
  ),
};
