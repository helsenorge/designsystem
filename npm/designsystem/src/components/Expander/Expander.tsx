import React, { useRef } from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';
import { useExpand } from '../../hooks/useExpand';
import { useHover } from '../../hooks/useHover';
import { useSize } from '../../hooks/useSize';
import { PaletteNames } from '../../theme/palette';
import Button from '../Button';
import Icon, { IconSize, SvgIcon } from '../Icon';
import ChevronDown from '../Icons/ChevronDown';
import ChevronUp from '../Icons/ChevronUp';
import { IconName } from '../Icons/IconNames';
import LazyIcon from '../LazyIcon';

import styles from './styles.module.scss';

export enum ExpanderSize {
  small = 'small',
  large = 'large',
}

export type ExpanderColors = Extract<PaletteNames, 'banana' | 'blueberry' | 'cherry' | 'kiwi' | 'neutral' | 'plum' | 'white'>;

export interface ExpanderProps {
  /** Sets the trigger title */
  title: string;
  /** Sets the expanded content */
  children?: React.ReactNode;
  /** Sets classnames on the content area */
  contentClassNames?: string;
  /** Sets the size of the expander. Default: ExpanderSize.small */
  size?: ExpanderSize;
  /** Sets the background of the expander. Requires size=ExpanderSize.large. */
  color?: ExpanderColors;
  /** Adds an icon to the expander trigger. Requires size=ExpanderSize.large. */
  svgIcon?: SvgIcon | IconName;
  /** Opens or closes the expander */
  expanded?: boolean;
  /** Removes border to the left of the content. Requires size=ExpanderSize.small. */
  noNestedLine?: boolean;
  /** Stick expander trigger to top of screen while scrolling down */
  sticky?: boolean;
  /** Called when expander is open/closed. */
  onExpand?: (isExpanded: boolean) => void;
  /** Whether to render children when closed (in which case they are hidden with CSS). Default: false */
  renderChildrenWhenClosed?: boolean;
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
  /** Overrides the default z-index of the expander header */
  zIndex?: number;
}

const Expander: React.FC<ExpanderProps> = props => {
  const {
    title,
    children,
    size = ExpanderSize.small,
    color,
    contentClassNames,
    svgIcon: icon,
    expanded = false,
    noNestedLine = false,
    sticky = false,
    testId,
    onExpand,
    renderChildrenWhenClosed = false,
    zIndex,
  } = props;
  const [isExpanded, setIsExpanded] = useExpand(expanded, onExpand);
  const expanderRef = useRef<HTMLDivElement>(null);
  const triggerRef = useRef<HTMLButtonElement>(null);
  const { isHovered } = useHover(triggerRef);
  const contentSize = useSize(expanderRef);
  const triggerSize = useSize(triggerRef);

  const isSticky = sticky && isExpanded;

  const renderChevron = (align: 'left' | 'right'): React.ReactNode => (
    <span className={classNames(styles['expander__icon'], styles[`expander__icon--${align}`])}>
      <Icon svgIcon={isExpanded ? ChevronUp : ChevronDown} size={IconSize.XSmall} isHovered={isHovered} />
    </span>
  );

  const triggerClassName = classNames(
    styles['expander__trigger'],
    size === ExpanderSize.large && styles[`expander__trigger--${size}`],
    size === ExpanderSize.large && styles[`expander__trigger--${color || 'neutral'}`],
    size === ExpanderSize.large && icon && styles['expander__trigger--icon'],
    isExpanded && styles['expander__trigger--expanded'],
    isSticky && styles['expander__trigger--sticky']
  );

  const renderTrigger = (): React.ReactNode => (
    <button
      type="button"
      className={triggerClassName}
      style={{
        zIndex: isHovered || isSticky ? zIndex : undefined,
        width: isSticky && contentSize?.width ? `${contentSize.width}px` : undefined,
      }}
      aria-expanded={isExpanded}
      ref={triggerRef}
      onClick={(): void => setIsExpanded(!isExpanded)}
      data-testid={testId}
      data-analyticsid={AnalyticsId.Expander}
    >
      {size === ExpanderSize.small && renderChevron('left')}
      {icon && (
        <span className={classNames(styles['expander__icon'], styles['expander__icon--left'])}>
          {typeof icon === 'string' ? (
            <LazyIcon iconName={icon} size={IconSize.XSmall} isHovered={isHovered} />
          ) : (
            <Icon svgIcon={icon} size={IconSize.XSmall} isHovered={isHovered} />
          )}
        </span>
      )}
      {title}
      {size === ExpanderSize.large && renderChevron('right')}
    </button>
  );

  const buttonClassName = classNames(styles['expander__button'], isExpanded && styles['expander__button--expanded']);

  const renderButton = (): React.ReactNode => (
    <div
      style={{
        width: isSticky && triggerSize?.width ? `${triggerSize?.width}px` : undefined,
        height: isSticky && triggerSize?.height ? `${triggerSize?.height}px` : undefined,
      }}
      className={classNames({
        [styles['expander__button-container--sticky']]: isSticky,
      })}
    >
      <Button
        variant="borderless"
        textClassName={styles['expander__button__text']}
        className={buttonClassName}
        aria-expanded={isExpanded}
        ref={triggerRef}
        onClick={(): void => setIsExpanded(!isExpanded)}
        testId={testId}
        data-analyticsid={AnalyticsId.Expander}
      >
        <Icon svgIcon={isExpanded ? ChevronUp : ChevronDown} size={IconSize.XSmall} />
        {title}
      </Button>
    </div>
  );

  const renderContent = (): React.ReactNode => {
    if (!renderChildrenWhenClosed && !isExpanded) {
      return null;
    }

    const contentClassName = classNames(
      styles['expander__content'],
      styles[`expander__content--${size}`],
      size === ExpanderSize.large && styles[`expander__content--${color || 'neutral'}`],
      size === ExpanderSize.large && icon && styles['expander__content--icon'],
      isExpanded && styles['expander__content--expanded'],
      size === ExpanderSize.small && !noNestedLine && styles['expander__content--nested-line'],
      { [styles['expander__content--sticky']]: isSticky },
      contentClassNames
    );

    return <div className={contentClassName}>{children}</div>;
  };

  return (
    <div className={styles['expander']} ref={expanderRef}>
      {size === ExpanderSize.large ? renderTrigger() : renderButton()}
      {renderContent()}
    </div>
  );
};

export default Expander;
