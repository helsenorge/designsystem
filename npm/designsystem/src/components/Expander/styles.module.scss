@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';
@import '../../scss/supernova/styles/colors';

.expander {
  position: relative;
  width: 100%;

  &:focus-within {
    z-index: 1;
  }

  @media print {
    border: 1px solid $black;
  }

  &__icon {
    display: flex;
    align-items: center;

    &--left {
      margin-right: getSpacer(2xs);

      @media (min-width: map.get($grid-breakpoints, md)) {
        margin-right: getSpacer(s);
      }
    }

    &--right {
      margin-left: auto;
      padding-left: getSpacer(2xs);
    }
  }

  &__button,
  &__trigger {
    &--sticky {
      position: fixed;
      top: 0;
    }

    &--absolute {
      position: absolute;
      bottom: 0;
    }
  }

  &__button {
    position: relative;
    padding-right: getSpacer(s);
    text-align: left;

    &:hover {
      background-color: var(--color-action-background-transparent-onlight-hover);
    }

    &--expanded {
      background-color: var(--color-action-background-transparent-onlight-hover);
    }
  }

  &__trigger {
    $trigger: &;

    position: relative;
    display: flex;
    align-items: center;
    border: none;
    outline: none;
    font-size: $font-size-sm;
    line-height: $lineheight-size-sm;
    font-weight: 600;
    text-align: left;

    @media (min-width: map.get($grid-breakpoints, md)) {
      font-size: $font-size-md;
      line-height: $lineheight-size-md;
    }

    &--large {
      width: 100%;
      padding: getSpacer(xs) getSpacer(2xs) getSpacer(xs) getSpacer(s);
      cursor: pointer;

      @media (min-width: map.get($grid-breakpoints, md)) {
        padding: getSpacer(xs) getSpacer(s) getSpacer(xs) getSpacer(l);
      }
    }

    &--icon {
      padding-left: getSpacer(2xs);

      @media (min-width: map.get($grid-breakpoints, md)) {
        padding-left: getSpacer(s);
      }
    }

    $colors: 'neutral', 'blueberry', 'cherry', 'kiwi', 'banana', 'plum';

    @each $color in $colors {
      &--#{$color} {
        background-color: map.get($palette-map, #{$color}50);

        &:hover {
          background-color: map.get($palette-map, #{$color}100);
        }

        &:focus {
          box-shadow:
            0 0 0 1.5px $neutral600,
            inset 0 0 0 1.5px $neutral600;
        }
        &#{$trigger}--expanded {
          background-color: map.get($palette-map, #{$color}100);

          &:hover {
            background-color: map.get($palette-map, #{$color}200);
          }
        }
      }
    }

    // Hvit bakgrunn har egne regler for bakgrunnsfarge, hover og fokus
    &--white {
      background-color: $white;

      &:hover {
        background-color: $neutral100;
      }

      &:focus {
        box-shadow:
          0 0 0 1.5px $neutral600,
          inset 0 0 0 1.5px $neutral600;
      }
      &#{$trigger}--expanded {
        box-shadow: inset 0 0 0 6px $neutral100;

        &:focus {
          box-shadow:
            0 0 0 1.5px $neutral600,
            inset 0 0 0 1.5px $neutral600,
            inset 0 0 0 6px $neutral100;
        }
      }
    }
  }

  &__content {
    $content: &;

    display: none;
    width: 100%;

    @media print {
      display: block;
    }

    &--expanded {
      display: block;
    }

    &--small {
      margin-top: getSpacer(s);
    }

    &--nested-line {
      padding-left: getSpacer(s);
      border-left: solid 3px $neutral200;

      @media (min-width: map.get($grid-breakpoints, md)) {
        margin-left: calc(38px / 2 - 1px); // Halvparten av chevron-ikonets bredde minus halve bredden på border
        padding-left: calc(
          38px / 2 + 12px - 2px
        ); // Halvparten av chevron-ikonets bredde, pluss ikonets margin-right på tablet, minus bredden på border
      }
    }

    &--large {
      padding: getSpacer(s) getSpacer(m) getSpacer(l) getSpacer(s);

      @media (min-width: map.get($grid-breakpoints, md)) {
        padding: getSpacer(m) getSpacer(l) getSpacer(xl) getSpacer(l);
      }
      &#{$content}--icon {
        @media (min-width: map.get($grid-breakpoints, md)) {
          padding-left: 70px;
        }
      }
    }

    &--neutral {
      background-color: $neutral50;
    }

    &--blueberry {
      background-color: $blueberry50;
    }

    &--cherry {
      background-color: $cherry50;
    }

    &--kiwi {
      background-color: $kiwi50;
    }

    &--banana {
      background-color: $banana50;
    }

    &--plum {
      background-color: $plum50;
    }

    &--white {
      background-color: $white;
    }
  }
}
