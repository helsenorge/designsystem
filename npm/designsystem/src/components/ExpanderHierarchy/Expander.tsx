import React from 'react';

import classNames from 'classnames';

import ExpanderButton from './ExpanderButton';
import ExpanderHierarchy, { ExpanderHierarchyCommonProps, ExpanderHierarchyProps } from './ExpanderHierarchy';
import { getHeadingTag } from './utils';
import { isComponent } from '../../utils/component';

import styles from './expander.module.scss';

export interface ExpanderProps extends ExpanderHierarchyCommonProps {
  /** Sets the trigger title */
  title: string;
  /** Opens or closes the expander */
  expanded?: boolean;
  /** Called when expander is open/closed. */
  onExpand?: (isExpanded: boolean) => void;
  /** Sets the expanded content */
  children?: React.ReactNode;
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
  /** Adds custom classes to the element. */
  className?: string;
}

export type ExpanderType = React.FC<ExpanderProps>;

const Expander: ExpanderType = ({
  title,
  htmlMarkup = 'h2',
  level = 1,
  print,
  expanded = false,
  onExpand,
  children,
  testId,
  className,
}: ExpanderProps) => {
  const contentClasses = classNames(
    styles.expander__content,
    (level === 1 || level === 2) && styles[`expander__content--${level}`],
    level > 2 && styles[`expander__content--3-and-lower`]
  );

  return (
    <li className={classNames(styles.expander, className)}>
      <ExpanderButton htmlMarkup={htmlMarkup} level={level} print={print} expanded={expanded} onExpand={onExpand} testId={testId}>
        {title}
      </ExpanderButton>
      <div className={contentClasses}>
        {React.Children.map(children, child =>
          isComponent<ExpanderHierarchyProps>(child, ExpanderHierarchy)
            ? React.cloneElement(child, { htmlMarkup: getHeadingTag(htmlMarkup), level: level + 1, print })
            : child
        )}
      </div>
    </li>
  );
};

export default Expander;
