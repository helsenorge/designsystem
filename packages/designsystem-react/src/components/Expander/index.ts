import Expander from './Expander';
import ExpanderList from './ExpanderList';
export {Expander, ExpanderList};
