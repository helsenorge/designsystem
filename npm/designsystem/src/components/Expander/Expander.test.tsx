import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Expander from './Expander';
import Calendar from '../Icons/Calendar';
import { isExportDeclaration } from 'typescript';

describe('Gitt at Expander skal vises vanlig', (): void => {
  describe('Når Expanderen vises', (): void => {
    test('Så kan man klikke for å ekspandere innholdet', (): void => {
      render(
        <Expander title={'Knapp'}>
          <h1>Innhold i expander</h1>
        </Expander>
      );
      const expander = screen.getByRole('button', { name: 'Knapp' });
      expect(expander).toHaveAttribute('aria-expanded', 'false');

      userEvent.click(expander);
      expect(expander).toHaveAttribute('aria-expanded', 'true');
      expect(screen.getByRole('heading', { name: 'Innhold i expander' })).toBeInTheDocument();
    });
    test('Så har den riktig analyticsid', (): void => {
      render(<Expander title={'Knapp'} testId={'knapp'}></Expander>);

      const expander = screen.getByTestId('knapp');

      expect(expander).toHaveAttribute('data-analyticsid', 'expander');
    });
  });
});
describe('Gitt at Expander skal vises som ekspandert', (): void => {
  describe('Når Expanderen vises', (): void => {
    test('Så vises innholdet ekspandert fra starten av', (): void => {
      render(
        <Expander title={'Knapp'} expanded>
          <h1>Innhold i expander</h1>
        </Expander>
      );

      const expander = screen.getByRole('button', { name: 'Knapp' });
      expect(expander).toHaveAttribute('aria-expanded', 'true');

      const content = screen.getByRole('heading', { name: 'Innhold i expander' });
      expect(content).toBeInTheDocument();
    });
  });
});
describe('Gitt at Expander skal vises som large, med farge og ikon', (): void => {
  describe('Når Expanderen vises', (): void => {
    test('Så rendres expanderen som den skal', (): void => {
      const { container } = render(
        <Expander title={'Knapp'} size="large" color="blueberry" svgIcon={Calendar}>
          <h1>Innhold i expander</h1>
        </Expander>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
describe('Gitt at Expander har onExpand-callback', (): void => {
  describe('Når man klikker på expanderen to ganger', (): void => {
    test('Så kalles callback først med true og så med false', (): void => {
      const handleExpand = jest.fn();
      render(<Expander title={'Knapp'} testId={'knapp'} onExpand={handleExpand}></Expander>);

      const expander = screen.getByTestId('knapp');

      userEvent.click(expander);
      userEvent.click(expander);

      expect(handleExpand).toHaveBeenCalledTimes(2);
      expect(handleExpand).toHaveBeenNthCalledWith(1, true);
      expect(handleExpand).toHaveBeenNthCalledWith(2, false);
    });
  });
});
describe('Gitt at Expander har onExpand-callback og expanded satt til true', (): void => {
  describe('Når man klikker på expanderen én gang', (): void => {
    test('Så kalles callback først med true og så med false', (): void => {
      const handleExpand = jest.fn();
      render(<Expander title={'Knapp'} testId={'knapp'} onExpand={handleExpand} expanded></Expander>);

      const expander = screen.getByTestId('knapp');

      userEvent.click(expander);

      expect(handleExpand).toHaveBeenCalledTimes(2);
      expect(handleExpand).toHaveBeenNthCalledWith(1, true);
      expect(handleExpand).toHaveBeenNthCalledWith(2, false);
    });
  });
});
describe('Gitt at Expander vises med et fokuserbart element', (): void => {
  describe('Når Expanderen vises', (): void => {
    test('Så kan man tabbe til elementet etter at ekspanderen er åpnet', (): void => {
      render(
        <Expander title={'Knapp'}>
          <button>En knapp til</button>
        </Expander>
      );

      const expander = screen.getByRole('button', { name: 'Knapp' });

      userEvent.click(expander);
      userEvent.tab();
      const focusableButton = screen.getByRole('button', { name: 'En knapp til' });
      expect(focusableButton).toHaveFocus();
    });
  });
});
describe('Gitt at Expander vises react-element som title', (): void => {
  describe('Når Expanderen vises', (): void => {
    test('Så fungerer Expanderen slik den skal', (): void => {
      render(
        <Expander title={<span data-testid="tittel">Tittel</span>}>
          <button>En knapp til</button>
        </Expander>
      );

      const title = screen.getByTestId('tittel');
      expect(title).toBeVisible();

      userEvent.click(title);
      const expander = screen.getByRole('button', { name: 'Tittel' });
      expect(expander).toHaveAttribute('aria-expanded', 'true');
    });
  });
});
