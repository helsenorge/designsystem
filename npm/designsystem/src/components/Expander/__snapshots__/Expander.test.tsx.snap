// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at Expander skal vises som large, med farge og ikon > NÃ¥r Expanderen vises > SÃ¥ rendres expanderen som den skal 1`] = `
<div>
  <div
    class="expander"
  >
    <button
      aria-expanded="false"
      class="expander__trigger expander__trigger--large expander__trigger--blueberry expander__trigger--icon"
      data-analyticsid="expander"
      type="button"
    >
      <span
        class="expander__icon expander__icon--left"
      >
        <svg
          aria-hidden="true"
          class="hnds-style-icon"
          color="black"
          data-analyticsid="icon"
          fill="black"
          focusable="false"
          height="38"
          role="presentation"
          style="min-width: 38px; min-height: 38px;"
          viewBox="0 0 48 48"
          width="38"
        >
          <path
            d="M38.887 16.71H8.6v-6.197h3.862v1.17c0 1.223.994 2.216 2.216 2.216a2.218 2.218 0 002.215-2.215v-1.171h13.582v1.17c0 1.223.994 2.216 2.215 2.216a2.218 2.218 0 002.216-2.215v-1.171h3.982v6.197zM8.6 34.46h30.288V17.974H8.6V34.46zm5.125-26.17a.953.953 0 011.905 0v3.394a.953.953 0 01-1.905 0V8.289zm18.013 0a.953.953 0 011.905 0v3.394a.953.953 0 01-1.905 0V8.289zm3.168.96v-.96a2.218 2.218 0 00-2.216-2.216 2.218 2.218 0 00-2.215 2.215v.96H16.892v-.96a2.218 2.218 0 00-2.215-2.215 2.218 2.218 0 00-2.216 2.215v.96H7.336v26.475h32.815V9.249h-5.246zM27.67 28.292a1.508 1.508 0 100-3.016 1.508 1.508 0 000 3.016z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      Knapp
      <span
        class="expander__icon expander__icon--right"
      >
        <svg
          aria-hidden="true"
          class="hnds-style-icon"
          color="black"
          data-analyticsid="icon"
          fill="black"
          focusable="false"
          height="38"
          role="presentation"
          style="min-width: 38px; min-height: 38px;"
          viewBox="0 0 48 48"
          width="38"
        >
          <path
            d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
